README.md
# fastapi-prometheus-exporter
[![codecov](https://codecov.io/gh/yibuma/fastapi-prometheus-exporter/graph/badge.svg?token=QDbwB7a9kP)](https://codecov.io/gh/yibuma/fastapi-prometheus-exporter)  
FastAPI Prometheus Exporter is a simple Prometheus exporter for FastAPI applications. It provides a set of metrics by default and allows you to add your own metrics as well.

## Installation
```bash
pip install fastapi-prometheus-exporter
```
or
```bash
poetry add fastapi-prometheus-exporter
```

## Usage
```python
from fastapi import FastAPI
from fastapi_prometheus_exporter import PrometheusExporterMiddleware
app = FastAPI()

PrometheusExporterMiddleware.setup(
    app=app,
    metrics_path="/metrics",
    # ignore_paths=["/healthz", "/metrics"],
)
```
### Parameters
- `app`: FastAPI application instance
- `ignore_paths`: List of paths to ignore. Default: `[]`
- `metrics_path`: Path to expose metrics. Default: `/metrics`
- `metrics_registry`: Prometheus registry to use. Default: `prometheus_client.CollectorRegistry`
- `request_latency_name`: Name of the request latency metric. Default: `http_request_duration_seconds`
- `request_latency_doc`: Description of the request latency metric. Default: `HTTP Request Latency`
- `request_latency_buckets`: Buckets for the request latency metric. Default: `prometheus_client.Histogram.DEFAULT_BUCKETS`
- `request_latency_labels`: Labels for the request latency metric. Default: `["method", "endpoint", "http_status"]`
- `request_count_name`: Name of the request count metric. Default: `http_requests_total`
- `request_count_doc`: Description of the request count metric. Default: `Total HTTP Requests`
- `request_count_labels`: Labels for the request count metric. Default: `["method", "endpoint", "http_status"]`

## Metrics
### Default Metrics
- `http_request_duration_seconds`: HTTP request latency in seconds
- `http_requests_total`: Total HTTP requests


CHANGELOG.md
# Changelog

## [0.2.2](https://github.com/yibuma/fastapi-prometheus-exporter/compare/v0.2.1...v0.2.2) (2023-12-04)


### Bug Fixes

* update starlette.exceptions to fastapi.HTTPException ([d00a70c](https://github.com/yibuma/fastapi-prometheus-exporter/commit/d00a70c13dc5fe89838e538f5f19e279238c9b6f))


### Documentation

* add README.md ([ad71514](https://github.com/yibuma/fastapi-prometheus-exporter/commit/ad715140063f2144b0f6508896a2d01abecf9e43))
* README add codecov badge ([846923e](https://github.com/yibuma/fastapi-prometheus-exporter/commit/846923edb296a639aea115b10f6c3230a7dfc9ad))

## [0.2.1](https://github.com/yibuma/fastapi-prometheus-exporter/compare/v0.2.0...v0.2.1) (2023-11-29)


### Bug Fixes

* fix reportMissingTypeStubs bug ([8b4e817](https://github.com/yibuma/fastapi-prometheus-exporter/commit/8b4e8178a14620c98d04a4848d98770c0ef79c48))

## [0.2.0](https://github.com/yibuma/fastapi-prometheus-exporter/compare/v0.1.0...v0.2.0) (2023-11-29)


### Features

* add prometheus exporter ([62f0a1b](https://github.com/yibuma/fastapi-prometheus-exporter/commit/62f0a1bfe684cd053e4c5a12851c6702b4abc526))


tests/test_prometheus.py
import unittest

from fastapi import FastAPI, HTTPException
from fastapi.testclient import TestClient

from fastapi_prometheus_exporter.prometheus import PrometheusExporterMiddleware

null_context = """# HELP http_requests_total Total HTTP Requests
# TYPE http_requests_total counter
# HELP http_request_duration_seconds HTTP Request Latency
# TYPE http_request_duration_seconds histogram
"""


def raise_(ex: Exception):
    raise ex


class TestPrometheus(unittest.TestCase):
    def setUp(self) -> None:
        app = FastAPI()
        app.get("/test")(lambda: {"status": "ok"})
        app.get("/ignore")(lambda: {"status": "ok"})
        app.get("/error")(lambda: 1 / 0)
        app.get("/exception")(lambda: raise_(HTTPException(500)))
        PrometheusExporterMiddleware.setup(
            app=app,
            ignore_paths=["/ignore"],
            metrics_path="/metrics",
        )
        client = TestClient(app)
        self.client = client

    def test_null_metrics(self):
        response = self.client.get("/metrics")
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.text, null_context)

    def test_ignore(self):
        response = self.client.get("/ignore")
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.json(), {"status": "ok"})
        response = self.client.get("/metrics")
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.text, null_context)

    def test_test(self):
        response = self.client.get("/test")
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.json(), {"status": "ok"})
        response = self.client.get("/metrics")
        self.assertEqual(response.status_code, 200)
        self.assertGreater(len(response.text), len(null_context))
        self.assertGreater(
            response.text.find(
                'http_requests_total{endpoint="/test",http_status="200",method="GET"} 1.0'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_requests_created{endpoint="/test",http_status="200",method="GET"}'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_request_duration_seconds_bucket{endpoint="/test",http_status="200",le="+Inf",method="GET"} 1.0'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_request_duration_seconds_count{endpoint="/test",http_status="200",method="GET"} 1.0'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_request_duration_seconds_sum{endpoint="/test",http_status="200",method="GET"}'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_request_duration_seconds_created{endpoint="/test",http_status="200",method="GET"} '
            ),
            0,
        )

    def test_404(self):
        response = self.client.get("/not-found")
        self.assertEqual(response.status_code, 404)
        response = self.client.get("/metrics")
        self.assertEqual(response.status_code, 200)
        self.assertGreater(len(response.text), len(null_context))
        self.assertGreater(
            response.text.find(
                'http_requests_total{endpoint="/not-found",http_status="404",method="GET"} 1.0'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_requests_created{endpoint="/not-found",http_status="404",method="GET"}'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_request_duration_seconds_bucket{endpoint="/not-found",http_status="404",le="+Inf",method="GET"} 1.0'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_request_duration_seconds_count{endpoint="/not-found",http_status="404",method="GET"} 1.0'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_request_duration_seconds_sum{endpoint="/not-found",http_status="404",method="GET"}'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_request_duration_seconds_created{endpoint="/not-found",http_status="404",method="GET"} '
            ),
            0,
        )

    def test_error(self):
        with self.assertRaises(ZeroDivisionError):
            response = self.client.get("/error")
            self.assertEqual(response.status_code, 500)
        response = self.client.get("/metrics")
        self.assertEqual(response.status_code, 200)
        self.assertGreater(len(response.text), len(null_context))
        self.assertGreater(
            response.text.find(
                'http_requests_total{endpoint="/error",http_status="500",method="GET"} 1.0'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_requests_created{endpoint="/error",http_status="500",method="GET"}'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_request_duration_seconds_bucket{endpoint="/error",http_status="500",le="+Inf",method="GET"} 1.0'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_request_duration_seconds_count{endpoint="/error",http_status="500",method="GET"} 1.0'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_request_duration_seconds_sum{endpoint="/error",http_status="500",method="GET"}'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_request_duration_seconds_created{endpoint="/error",http_status="500",method="GET"} '
            ),
            0,
        )

    def test_exception(self):
        response = self.client.get("/exception")
        self.assertEqual(response.status_code, 500)
        response = self.client.get("/metrics")
        self.assertEqual(response.status_code, 200)
        self.assertGreater(len(response.text), len(null_context))
        self.assertGreater(
            response.text.find(
                'http_requests_total{endpoint="/exception",http_status="500",method="GET"} 1.0'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_requests_created{endpoint="/exception",http_status="500",method="GET"}'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_request_duration_seconds_bucket{endpoint="/exception",http_status="500",le="+Inf",method="GET"} 1.0'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_request_duration_seconds_count{endpoint="/exception",http_status="500",method="GET"} 1.0'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_request_duration_seconds_sum{endpoint="/exception",http_status="500",method="GET"}'
            ),
            0,
        )
        self.assertGreater(
            response.text.find(
                'http_request_duration_seconds_created{endpoint="/exception",http_status="500",method="GET"} '
            ),
            0,
        )


fastapi_prometheus_exporter/prometheus.py
import timeit
import typing

from fastapi import FastAPI, Request, Response, HTTPException
from fastapi.responses import PlainTextResponse

from prometheus_client import (
    Counter,
    Histogram,
    CollectorRegistry,
    CONTENT_TYPE_LATEST,
    generate_latest,
)


class PrometheusExporterMiddleware:
    @classmethod
    def setup(
        cls,
        app: FastAPI,
        ignore_paths: list[str] | None = None,
        metrics_path: str = "/metrics",
        metrics_registry: CollectorRegistry | None = None,
        request_latency_name: str = "http_request_duration_seconds",
        request_latency_doc: str = "HTTP Request Latency",
        request_latency_buckets: typing.Sequence[
            float | str
        ] = Histogram.DEFAULT_BUCKETS,
        request_latency_labels: list[str] = ["method", "endpoint", "http_status"],
        request_count_name: str = "http_requests_total",
        request_count_doc: str = "Total HTTP Requests",
        request_count_labels: list[str] = ["method", "endpoint", "http_status"],
    ):
        if metrics_registry is None:
            metrics_registry = CollectorRegistry()
        request_count = Counter(
            request_count_name,
            request_count_doc,
            request_count_labels,
            registry=metrics_registry,
        )
        request_latency = Histogram(
            request_latency_name,
            request_latency_doc,
            request_latency_labels,
            buckets=request_latency_buckets,
            registry=metrics_registry,
        )
        middleware = cls(
            app,
            request_latency,
            request_count,
            metrics_registry,
            ignore_paths,
            metrics_path,
        )
        middleware.register()

    def __init__(
        self,
        app: FastAPI,
        request_latency: Histogram,
        request_count: Counter,
        registry: CollectorRegistry,
        ignore_paths: list[str] | None,
        metrics_path: str,
    ):
        self.app = app
        self.request_latency = request_latency
        self.request_count = request_count
        self.registry = registry
        self.ignore_paths = ignore_paths
        self.metrics_path = metrics_path

    def register(self):
        self.app.get(self.metrics_path, response_class=PlainTextResponse)(
            self.prometheus_output
        )
        self.app.middleware("http")(self.prometheus_middleware)

    async def prometheus_output(self):
        return PlainTextResponse(
            content=generate_latest(self.registry),
            media_type=CONTENT_TYPE_LATEST,
        )

    async def prometheus_middleware(
        self,
        request: Request,
        call_next: typing.Callable[[Request], typing.Awaitable[Response]],
    ):
        endpoint = request.url.path
        if endpoint == self.metrics_path or (
            self.ignore_paths and endpoint in self.ignore_paths
        ):
            return await call_next(request)

        route = request.scope.get("route")
        if route is not None:
            if route.path in self.ignore_paths:
                return await call_next(request)
            endpoint = route.path

        status_code = 200
        start_time = timeit.default_timer()
        try:
            response = await call_next(request)
            status_code = response.status_code
            return response
        except HTTPException as e:
            status_code = e.status_code
            raise e
        except Exception as e:
            status_code = 500
            raise e
        finally:
            request_latency = timeit.default_timer() - start_time
            self.request_latency.labels(
                method=request.method,
                endpoint=endpoint,
                http_status=status_code,
            ).observe(request_latency)
            self.request_count.labels(
                method=request.method,
                endpoint=endpoint,
                http_status=status_code,
            ).inc()
